Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs

```diff
@@ -0,0 +1,107 @@
+// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections;
+using System.Data;
+using System.Data.SqlClient;
+using System.Linq;
+using Microsoft.SqlServer.Server;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Model;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Parameters;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.Parameters;
+
+/// <summary>
+/// Handles parameter values that have to be represented as a table-valued parameter.
+/// </summary>
+public class SqlTableValuedDataParameterDefinition : IDataParameterDefinition
+{
+  public const string DistinctCollectionTableTypeNameSuffix = "_distinct";
+  public bool IsDistinct { get; }
+  public IStorageTypeInformation StorageTypeInformation { get; }
+
+  public SqlTableValuedDataParameterDefinition (IStorageTypeInformation storageTypeInformation, bool isDistinct)
+  {
+    ArgumentUtility.CheckNotNull(nameof(storageTypeInformation), storageTypeInformation);
+
+    IsDistinct = isDistinct;
+    StorageTypeInformation = storageTypeInformation;
+  }
+
+  /// <summary>
+  /// Creates an <see cref="SqlTableValuedParameterValue"/> from a <paramref name="value"/> that is a collection of items whose type is compatible with the
+  /// <see cref="StorageTypeInformation"/>'s <see cref="IStorageTypeInformation.DotNetType"/>.
+  /// </summary>
+  public object GetParameterValue (object? value)
+  {
+    var enumerable = ArgumentUtility.CheckType<IEnumerable?>(nameof(value), value);
+    if (enumerable == null)
+      return DBNull.Value;
+
+    var convertedValues = new ArrayList();
+    foreach (var item in enumerable)
+    {
+      if (item != null && !StorageTypeInformation.DotNetType.IsInstanceOfType(item))
+        throw ArgumentUtility.CreateArgumentItemTypeException(nameof(value), convertedValues.Count, StorageTypeInformation.DotNetType, item.GetType());
+
+      var convertedValue = StorageTypeInformation.ConvertToStorageType(item);
+      convertedValues.Add(convertedValue);
+    }
```

Comment:
I don't think we should do a copy here. we can simply iterate and perform `tvpValue.AddRecord` instead.

We have use cases with thousands of rows in the parameter and doing two copies is more expensive than doing one copy.

Reply:
Do you mean we should only check the items for their type here, and then re-iterate the `enumerable` after we created the `SqlTableValuedParameterValue`? Or do we iterate only after creating the `SqlTableValuedParameterValue`, but still throw an `ArgumentItemTypeException` at the end of the method?